File: src/main/java/main/commands/slashcommands/commands/management/requestchannel/RequestChannelEditCommand.java
package main.commands.slashcommands.commands.management.requestchannel;

import lombok.extern.slf4j.Slf4j;
import main.constants.Permission;
import main.constants.RobertifyEmoji;
import main.utils.GeneralUtils;
import main.utils.RobertifyEmbedUtils;
import main.utils.component.interactions.AbstractSlashCommand;
import main.utils.json.requestchannel.RequestChannelConfig;
import main.utils.locale.LocaleManager;
import main.utils.locale.RobertifyLocaleMessage;
import net.dv8tion.jda.api.entities.Guild;
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent;
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent;
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException;
import net.dv8tion.jda.api.interactions.components.buttons.Button;
import net.dv8tion.jda.internal.utils.tuple.Pair;
import org.jetbrains.annotations.NotNull;
import org.jetbrains.annotations.Nullable;

import java.util.concurrent.TimeUnit;

@Slf4j
public class RequestChannelEditCommand extends AbstractSlashCommand {
    @Override
    protected void buildCommand() {
        setCommand(
                getBuilder()
                        .setName("requestchannel")
                        .setDescription("Configure the request channel for your server!")
                        .addSubCommands(
                                SubCommand.of(
                                        "setup",
                                        "Setup the request channel!"
                                ),
                                SubCommand.of(
                                        "edit",
                                        "Edit the request channel!"
                                )
                        )
                        .build()
        );
    }

    @Override
    public String getHelp() {
        return null;
    }

    protected void handleSetup(SlashCommandInteractionEvent event) {
        final var guild = event.getGuild();

        if (new RequestChannelConfig(guild).isChannelSet()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_ALREADY_SETUP).build())
                    .queue();
            return;
        }
        event.deferReply().queue();
        new RequestChannelCommand().createRequestChannel(event.getGuild())
                .thenAccept(channel -> {
                    try {
                        event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SETUP, Pair.of("{channel}", GeneralUtils.toMention(guild, channel.getChannelId(), GeneralUtils.Mentioner.CHANNEL))).build())
                                .queue();
                    } catch (InsufficientPermissionException e) {
                        if (e.getMessage().contains("MESSAGE_HISTORY"))
                            event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SETUP_2).build())
                                    .queue();
                        else log.error("Unexpected error", e);
                    }
                });
    }

    private void handleEdit(SlashCommandInteractionEvent event) {
        final var guild = event.getGuild();

        if (!GeneralUtils.hasPerms(guild, event.getMember(), Permission.ROBERTIFY_ADMIN)) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.INSUFFICIENT_PERMS, Pair.of("{permissions}", Permission.ROBERTIFY_ADMIN.name())).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        RequestChannelConfig requestChannelConfig = new RequestChannelConfig(guild);
        if (!requestChannelConfig.isChannelSet()) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_NOT_SET).build())
                    .queue();
            return;
        }

        event.deferReply().queue();

        final var localeManager = LocaleManager.getLocaleManager(guild);

        event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_EDIT_EMBED)
                        .build())
                .addActionRow(
                        Button.primary("togglerqchannel:previous:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_PREVIOUS)).withEmoji(RobertifyEmoji.PREVIOUS_EMOJI.getEmoji()),
                        Button.primary("togglerqchannel:rewind:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_REWIND)).withEmoji(RobertifyEmoji.REWIND_EMOJI.getEmoji()),
                        Button.primary("togglerqchannel:pnp:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_PLAY_AND_PAUSE)).withEmoji(RobertifyEmoji.PLAY_AND_PAUSE_EMOJI.getEmoji()),
                        Button.primary("togglerqchannel:stop:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_STOP)).withEmoji(RobertifyEmoji.STOP_EMOJI.getEmoji()),
                        Button.primary("togglerqchannel:skip:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SKIP)).withEmoji(RobertifyEmoji.END_EMOJI.getEmoji())
                )
                .addActionRow(
                        Button.secondary("togglerqchannel:favourite:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_FAVOURITE)).withEmoji(RobertifyEmoji.STOP_EMOJI.getEmoji()),
                        Button.secondary("togglerqchannel:loop:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_LOOP)).withEmoji(RobertifyEmoji.LOOP_EMOJI.getEmoji()),
                        Button.secondary("togglerqchannel:shuffle:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SHUFFLE)).withEmoji(RobertifyEmoji.SHUFFLE_EMOJI.getEmoji()),
                        Button.secondary("togglerqchannel:disconnect:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_DISCONNECT)).withEmoji(RobertifyEmoji.QUIT_EMOJI.getEmoji()),
                        Button.secondary("togglerqchannel:filters:" + event.getUser().getId(), localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_FILTERS)).withEmoji(RobertifyEmoji.FILTER_EMOJI.getEmoji())
                )
                .setEphemeral(true)
                .queue();
    }

    @Override
    public void onButtonInteraction(@NotNull ButtonInteractionEvent event) {
        if (!event.getButton().getId().startsWith("togglerqchannel:")) return;

        final var guild = event.getGuild();
        final var split = event.getButton().getId().split(":");
        if (!split[2].equals(event.getUser().getId())) {
            event.replyEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.GeneralMessages.NO_PERMS_BUTTON).build())
                    .setEphemeral(true)
                    .queue();
            return;
        }

        event.deferReply().queue();
        handleChannelButtonToggle(guild, split[1], event);
    }
    
    public void handleChannelButtonToggle(Guild guild, String button, @Nullable ButtonInteractionEvent event) {
        final LocaleManager localeManager = LocaleManager.getLocaleManager(guild);
        final var dedicatedChannelConfig = new RequestChannelConfig(guild);
        final var config = dedicatedChannelConfig.getConfig();
        final RequestChannelConfig.ChannelConfig.Field field;
        final String buttonName;
        
        switch (button) {
            case "previous" -> {
                field = RequestChannelConfig.ChannelConfig.Field.PREVIOUS;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_PREVIOUS);
            }
            case "rewind" -> {
                field = RequestChannelConfig.ChannelConfig.Field.REWIND;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_REWIND);
            }
            case "pnp" -> {
                field = RequestChannelConfig.ChannelConfig.Field.PLAY_PAUSE;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_PLAY_AND_PAUSE);
            }
            case "stop" -> {
                field = RequestChannelConfig.ChannelConfig.Field.STOP;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_STOP);
            }
            case "skip" -> {
                field = RequestChannelConfig.ChannelConfig.Field.SKIP;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SKIP);
            }
            case "favourite" -> {
                field = RequestChannelConfig.ChannelConfig.Field.FAVOURITE;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_FAVOURITE);
            }
            case "loop" -> {
                field = RequestChannelConfig.ChannelConfig.Field.LOOP;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_LOOP);
            }
            case "shuffle" -> {
                field = RequestChannelConfig.ChannelConfig.Field.SHUFFLE;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_SHUFFLE);
            }
            case "disconnect" -> {
                field = RequestChannelConfig.ChannelConfig.Field.DISCONNECT;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_DISCONNECT);
            }
            case "filters" -> {
                field = RequestChannelConfig.ChannelConfig.Field.FILTERS;
                buttonName = localeManager.getMessage(RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_FILTERS);
            }
            default -> throw new IllegalArgumentException("The button ID doesn't map to a case to be handled! ID: " + event.getButton().getId());
        }

        if (config.getState(field)) {
            config.setState(field, false);
            if (event != null)
                event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_BUTTON_TOGGLE,
                                Pair.of("{button}", buttonName),
                                Pair.of("{status}", localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.OFF_STATUS).toUpperCase())
                        ).build())
                        .queue(msg -> msg.delete().queueAfter(15, TimeUnit.SECONDS));
        } else {
            config.setState(field, true);
            if (event != null)
                event.getHook().sendMessageEmbeds(RobertifyEmbedUtils.embedMessage(guild, RobertifyLocaleMessage.DedicatedChannelMessages.DEDICATED_CHANNEL_BUTTON_TOGGLE,
                                Pair.of("{button}", buttonName),
                                Pair.of("{status}", localeManager.getMessage(RobertifyLocaleMessage.GeneralMessages.ON_STATUS).toUpperCase())
                        ).build())
                        .queue(msg -> msg.delete().queueAfter(15, TimeUnit.SECONDS));
        }

        dedicatedChannelConfig.updateButtons();
    }

    @Override
    public void onSlashCommandInteraction(@NotNull SlashCommandInteractionEvent event) {
        if (!checks(event)) return;

        switch (event.getSubcommandName()) {
            case "setup" -> handleSetup(event);
            case "edit" -> handleEdit(event);
        }
    }
}

File: src/main/kotlin/main/commands/slashcommands/management/requestchannel/RequestChannelEditCommand.kt
package main.commands.slashcommands.management.requestchannel

import dev.minn.jda.ktx.interactions.components.primary
import dev.minn.jda.ktx.interactions.components.secondary
import dev.minn.jda.ktx.messages.send
import dev.minn.jda.ktx.util.SLF4J
import kotlinx.coroutines.*
import main.audiohandlers.RobertifyAudioManager
import main.audiohandlers.loaders.MainAudioLoader.Companion.queueThenDelete
import main.constants.RobertifyEmoji
import main.constants.Toggle
import main.main.Robertify
import main.utils.GeneralUtils
import main.utils.GeneralUtils.toMention
import main.utils.RobertifyEmbedUtils
import main.utils.RobertifyEmbedUtils.Companion.replyEmbed
import main.utils.RobertifyEmbedUtils.Companion.sendEmbed
import main.utils.component.interactions.slashcommand.AbstractSlashCommand
import main.utils.component.interactions.slashcommand.models.SlashCommand
import main.utils.component.interactions.slashcommand.models.SubCommand
import main.utils.json.requestchannel.RequestChannelButton
import main.utils.json.requestchannel.RequestChannelConfig
import main.utils.json.requestchannel.RequestChannel
import main.utils.json.themes.ThemesConfig
import main.utils.json.toggles.TogglesConfig
import main.utils.locale.LocaleManager
import main.utils.locale.messages.DedicatedChannelMessages
import main.utils.locale.messages.GeneralMessages
import net.dv8tion.jda.api.EmbedBuilder
import net.dv8tion.jda.api.entities.Guild
import net.dv8tion.jda.api.events.interaction.command.SlashCommandInteractionEvent
import net.dv8tion.jda.api.events.interaction.component.ButtonInteractionEvent
import net.dv8tion.jda.api.exceptions.InsufficientPermissionException
import net.dv8tion.jda.api.interactions.components.ActionRow
import net.dv8tion.jda.api.sharding.ShardManager
import java.util.concurrent.TimeUnit
import java.util.concurrent.atomic.AtomicLong

class RequestChannelEditCommand : AbstractSlashCommand(
    SlashCommand(
        name = "requestchannel",
        description = "Configure the request channel for your server.",
        adminOnly = true,
        subcommands = listOf(
            SubCommand(
                name = "setup",
                description = "Setup the request channel."
            ),
            SubCommand(
                name = "edit",
                description = "Edit the request channel."
            )
        )
    )
) {

    companion object {
        @OptIn(DelicateCoroutinesApi::class)
        private val threadContext = newFixedThreadPoolContext(4, "Request-Channel-Worker")
        private val logger by SLF4J

        private const val EDIT_BUTTON_ID = "togglerqchannel:%s:%s"
    }

    override fun handle(event: SlashCommandInteractionEvent) = when (event.subcommandName) {
        "setup" -> handleSetup(event)
        else -> handleEdit(event)
    }

    internal fun handleSetup(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!

        if (RequestChannelConfig(guild).isChannelSet()) {
            event.replyEmbed {
                embed(DedicatedChannelMessages.DEDICATED_CHANNEL_ALREADY_SETUP)
            }.queue()
            return
        }

        event.deferReply().queue()
        val channel = createRequestChannel(guild)

        try {
            event.hook.sendEmbed(guild) {
                embed(
                    DedicatedChannelMessages.DEDICATED_CHANNEL_SETUP,
                    Pair("{channel}", channel.channelId.toMention(GeneralUtils.Mentioner.CHANNEL))
                )
            }.queue()
        } catch (e: InsufficientPermissionException) {
            if (e.message?.contains("MESSAGE_HISTORY") == true)
                event.hook.sendEmbed(guild) {
                    embed(DedicatedChannelMessages.DEDICATED_CHANNEL_SETUP_2)
                }.queue()
            else logger.error("Unexpected error", e)
        }
    }

    fun createRequestChannel(
        guild: Guild,
        shardManager: ShardManager = Robertify.shardManager
    ): RequestChannel {
        val job = run {
            val config = RequestChannelConfig(guild, shardManager)

            val textChannelId = AtomicLong()
            val localeManager = LocaleManager[guild]
            val theme = ThemesConfig(guild).getTheme()
            val embedBuilder = EmbedBuilder()
            embedBuilder.setColor(theme.color)
                .setTitle(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_NOTHING_PLAYING))
                .setImage(theme.idleBanner)

            val channel = guild.createTextChannel("robertify-requests").submit()
                .thenApply { channel ->
                    val manager = channel.manager

                    manager.setPosition(0).queue()
                    config.channelTopicUpdateRequest(channel)?.queue()

                    textChannelId.set(channel.idLong)
                    channel
                }
                .thenApply { channel ->
                    channel.sendMessage(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_QUEUE_NOTHING_PLAYING))
                        .setEmbeds(embedBuilder.build())
                        .submit()
                        .thenApply { message ->
                            config.setChannelAndMessage(textChannelId.get(), message.idLong)
                            config.buttonUpdateRequest(message).queue()
                            config.setOriginalAnnouncementToggle(TogglesConfig(guild).getToggle(Toggle.ANNOUNCE_MESSAGES))

                            try {
                                if (RobertifyAudioManager[guild].player.playingTrack != null)
                                    config.updateMessage()
                            } catch (_: UninitializedPropertyAccessException) {
                            }

                            RequestChannel(
                                channelId = config.getChannelId(),
                                messageId = config.getMessageId(),
                                config = config.config.getConfigJsonObject()
                            )
                        }.join()
                }.join()
            return@run channel
        }

        return job
    }

    fun deleteRequestChannel(guild: Guild, shardManager: ShardManager = Robertify.shardManager) {
        RequestChannelConfig(guild, shardManager).removeChannel()
    }

    private fun handleEdit(event: SlashCommandInteractionEvent) {
        val guild = event.guild!!
        val config = RequestChannelConfig(guild)

        if (!config.isChannelSet()) {
            event.replyEmbed {
                embed(DedicatedChannelMessages.DEDICATED_CHANNEL_NOT_SET)
            }.queue()
            return
        }

        event.deferReply().queue()
        val localeManager = LocaleManager[guild]
        val user = event.user

        event.hook.send(
            embeds = listOf(
                RobertifyEmbedUtils.embedMessage(
                    guild,
                    DedicatedChannelMessages.DEDICATED_CHANNEL_EDIT_EMBED
                ).build()
            ),
            components = listOf(
                ActionRow.of(
                    primary(
                        id = EDIT_BUTTON_ID.format("previous", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_PREVIOUS),
                        emoji = RobertifyEmoji.PREVIOUS_EMOJI.emoji
                    ),
                    primary(
                        id = EDIT_BUTTON_ID.format("rewind", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_REWIND),
                        emoji = RobertifyEmoji.REWIND_EMOJI.emoji
                    ),
                    primary(
                        id = EDIT_BUTTON_ID.format("pnp", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_PLAY_AND_PAUSE),
                        emoji = RobertifyEmoji.PLAY_AND_PAUSE_EMOJI.emoji
                    ),
                    primary(
                        id = EDIT_BUTTON_ID.format("stop", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_STOP),
                        emoji = RobertifyEmoji.STOP_EMOJI.emoji
                    ),
                    primary(
                        id = EDIT_BUTTON_ID.format("skip", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_SKIP),
                        emoji = RobertifyEmoji.END_EMOJI.emoji
                    ),
                ),
                ActionRow.of(
                    secondary(
                        id = EDIT_BUTTON_ID.format("favourite", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_FAVOURITE),
                        emoji = RobertifyEmoji.STAR_EMOJI.emoji
                    ),
                    secondary(
                        id = EDIT_BUTTON_ID.format("loop", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_LOOP),
                        emoji = RobertifyEmoji.LOOP_EMOJI.emoji
                    ),
                    secondary(
                        id = EDIT_BUTTON_ID.format("shuffle", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_SHUFFLE),
                        emoji = RobertifyEmoji.SHUFFLE_EMOJI.emoji
                    ),
                    secondary(
                        id = EDIT_BUTTON_ID.format("disconnect", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_DISCONNECT),
                        emoji = RobertifyEmoji.QUIT_EMOJI.emoji
                    ),
                    secondary(
                        id = EDIT_BUTTON_ID.format("filters", user.id),
                        label = localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_FILTERS),
                        emoji = RobertifyEmoji.STAR_EMOJI.emoji
                    ),
                )
            ),
            ephemeral = true
        ).queue()
    }

    override fun onButtonClick(event: ButtonInteractionEvent) {
        if (event.button.id?.startsWith(EDIT_BUTTON_ID.split(":")[0]) != true) return

        val guild = event.guild!!
        val (_, buttonName, allowedUserId) = event.button.id!!.split(":")
        if (allowedUserId != event.user.id) {
            event.replyEmbed {
                embed(GeneralMessages.NO_PERMS_BUTTON)
            }
                .setEphemeral(true)
                .queue()
            return
        }

        event.deferReply().queue()
        handleChannelButtonToggle(guild, listOf(buttonName), event)
    }

    fun handleChannelButtonToggle(
        guild: Guild,
        buttonNames: List<String>,
        event: ButtonInteractionEvent? = null,
        shardManager: ShardManager = Robertify.shardManager
    ) {
        val localeManager = LocaleManager[guild]
        val config = RequestChannelConfig(guild, shardManager)
        val subConfig = config.config
        val fields: MutableList<RequestChannelButton> = mutableListOf()
        val buttons: MutableList<String> = mutableListOf()

        buttonNames.forEach { buttonName ->
            when (buttonName) {
                "previous" -> {
                    fields.add(RequestChannelButton.PREVIOUS)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_PREVIOUS))
                }

                "rewind" -> {
                    fields.add(RequestChannelButton.REWIND)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_REWIND))
                }

                "pnp", "play_and_pause", "play_pause" -> {
                    fields.add(RequestChannelButton.PLAY_PAUSE)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_PLAY_AND_PAUSE))
                }

                "stop" -> {
                    fields.add(RequestChannelButton.STOP)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_STOP))
                }

                "skip" -> {
                    fields.add(RequestChannelButton.SKIP)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_SKIP))
                }

                "favourite" -> {
                    fields.add(RequestChannelButton.FAVOURITE)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_FAVOURITE))
                }

                "loop" -> {
                    fields.add(RequestChannelButton.LOOP)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_LOOP))
                }

                "shuffle" -> {
                    fields.add(RequestChannelButton.SHUFFLE)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_SHUFFLE))
                }

                "disconnect" -> {
                    fields.add(RequestChannelButton.DISCONNECT)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_DISCONNECT))
                }

                "filters" -> {
                    fields.add(RequestChannelButton.FILTERS)
                    buttons.add(localeManager.getMessage(DedicatedChannelMessages.DEDICATED_CHANNEL_FILTERS))

                }

                else -> throw IllegalArgumentException("The button ID \"${event?.button?.id!!}\" doesn't map to a case to be handled!")
            }
        }

        subConfig.toggleStates(fields)

        if (event != null)
            fields.forEachIndexed { i, field ->
                if (subConfig.getState(field)) {
                    event.hook.sendEmbed(guild) {
                        embed(
                            DedicatedChannelMessages.DEDICATED_CHANNEL_BUTTON_TOGGLE,
                            Pair("{button}", buttons[i]),
                            Pair("{status}", localeManager.getMessage(GeneralMessages.ON_STATUS))
                        )
                    }.queueThenDelete(time = 15, unit = TimeUnit.SECONDS)
                } else {
                    event.hook.sendEmbed(guild) {
                        embed(
                            DedicatedChannelMessages.DEDICATED_CHANNEL_BUTTON_TOGGLE,
                            Pair("{button}", buttons[i]),
                            Pair("{status}", localeManager.getMessage(GeneralMessages.OFF_STATUS))
                        )
                    }.queueThenDelete(time = 15, unit = TimeUnit.SECONDS)
                }
            }

        return config.updateButtons()
    }

}